### ISSUE TITLE
[Bug] `beartype.roar.BeartypeDoorHintViolation` deserialization error :panda_face: 

### ISSUE THREAD

**Issue by uriyasama** [beartype/beartype#443]
Hi!
This is a comparably small issue, but nonetheless an important one

Serializing and deserializing BeartypeDoorHintViolation fails on the following error:
`    def unpickle_exception(func, args, cause, tb, context=None, suppress_context=False, notes=None):
>       inst = func(*args)
E       TypeError: __init__() missing 1 required positional argument: 'culprits'`

Seems like serializing it returns a json with the field _culprits_weakref_and_repr instead of culprits.

**Comment by leycec** [comment]
**Ugh!** That definitely looks like badness. Would you happen to have a minimal working example (MWE)? If not, no worries. I have a reasonable idea of the (*...waitforit*) culprit. It looks like our `beartype.roar.BeartypeCallHintViolation` superclass *probably* needs to define a new pair of `__getstate__()` and `__setstate__()` dunder methods to manage `pickle` and `dill` (de)serialization across process boundaries now, huh?

I sigh, fam. I sigh so hard. :face_exhaling:

**Comment by leycec** [comment]
> ...your melodramatic responses are always a pleasure.

![](https://media1.tenor.com/m/O2H3NC5W5OcAAAAC/what-you-say-zerowung.gif)

> Beartype just wouldn't be the same without your brand of speech :)

**Awwwwwww.** Thanks so much for the kindest words. Tonight, we crush all the bugs for @uriyasama. :heart_eyes:

**Comment by leycec** [comment]
Keanu says: **"WOAH."**

Your example actually runs for me without error. I tried everything, @uriyasama. I swapped `dill` for `pickle`. *Still worked.* I started doing crazy "guaranteed to fail" stuff like forcing garbage collection by calling `gc.collect()` in between the calls to `dumps(e)` and `loads(serialized_exception)`. *Still works.* No idea why. Python hates this one simple trick.

Actually, I *do* have an idea why. For efficiency, Python **interns** (i.e., internalizes as globally cached objects) simple primitives like strings. How and when Python does so probably varies depending on platform and compiler options and *whatevahs.* Python is thus probably interning the `"hi"` string in your example on your Python but *not* my Python. This discrepancy introduces non-determinism that then causes my veiny forehead to bulge out like a loaf of Wonderbread.

Thankfully, I am a stubborn old man with a bald head. People like us never give up! I extrapolated a new example from your example that eliminates the interning and thus non-determinism by instantiating a user-defined pure-Python class. Since Python does *not* intern pure-Python objects, we expose the horrid truth for all to see:

```python
from pickle import dumps, loads
from beartype.door import die_if_unbearable
from beartype.roar import BeartypeDoorHintViolation
# from gc import collect

class TheHorridTruth(object): pass
the_awful_facts = TheHorridTruth()

try:
    die_if_unbearable(the_awful_facts, int)
except BeartypeDoorHintViolation as exception:
    exception_dumped = dumps(exception)
    exception_loaded = loads(exception_dumped)

    assert isinstance(exception, BeartypeDoorHintViolation)
    assert len(exception.culprits) == 1
    assert len(exception_loaded.culprits) == 1
    assert exception.culprits[0] == exception_loaded.culprits[0]
```

...which raises:

```python
Traceback (most recent call last):
  File "/home/leycec/tmp/mopy.py", line 11, in <module>
    die_if_unbearable(the_awful_facts, int)
    ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "/home/leycec/py/beartype/beartype/door/_func/doorcheck.py", line 126, in die_if_unbearable
    func_raiser(obj)  # pyright: ignore[reportUnboundVariable]
    ~~~~~~~~~~~^^^^^
  File "<@beartype(__beartype_checker_0) at 0x5574cbbcc160>", line 19, in __beartype_checker_0
beartype.roar.BeartypeDoorHintViolation: Die_if_unbearable() value <__main__.TheHorridTruth object at 0x7f99c0351e80> violates type hint <class 'int'>, as <class "__main__.TheHorridTruth"> <__main__.TheHorridTruth object at 0x7f99c0351e80> not instance of int.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/leycec/tmp/mopy.py", line 13, in <module>
    exception_dumped = dumps(exception)
TypeError: cannot pickle 'weakref.ReferenceType' object
```

![](https://media1.tenor.com/m/wn4IlOZj4cIAAAAd/solar-highway-great-success.gif)

**Comment by leycec** [comment]
**Resolved by d9e52d0.** Aaaaaaaaaaaaaand... that was sheer madness. :sweat:
